runtime/src/bank.rs

```diff
@@ -1907,85 +1958,114 @@ impl Bank {
-            let mut vote_account_changed = false;
-            let voters_account_pre_balance = vote_account.lamports;
-            let vote_state: VoteState = match StateMut::<VoteStateVersions>::state(vote_account) {
-                Ok(vote_state) => vote_state.convert_to_current(),
-                Err(err) => {
-                    debug!(
-                        "failed to deserialize vote account {}: {}",
-                        vote_pubkey, err
-                    );
-                    continue;
-                }
-            };
-            let commission = Some(vote_state.commission);
-
-            for (stake_pubkey, stake_account) in stake_group.iter_mut() {
-                // curry closure to add the contextual stake_pubkey
-                let mut reward_calc_tracer = reward_calc_tracer.as_mut().map(|outer| {
-                    let stake_pubkey = *stake_pubkey;
-                    // inner
-                    move |inner_event: &_| {
-                        outer(&RewardCalculationEvent::Staking(&stake_pubkey, inner_event))
-                    }
-                });
-                let redeemed = stake_state::redeem_rewards(
-                    rewarded_epoch,
-                    stake_account,
+        let point_value = PointValue { rewards, points };
+        let vote_account_rewards: DashMap<Pubkey, (AccountSharedData, u8, u64)> =
+            DashMap::with_capacity(stake_delegation_accounts.len());
+        let stake_delegation_iterator = stake_delegation_accounts.into_iter().flat_map(
+            |(
+                vote_pubkey,
+                StakeDelegationAccounts {
+                    vote_state,
                     vote_account,
-                    &vote_state,
-                    &point_value,
-                    Some(&stake_history),
-                    &mut reward_calc_tracer.as_mut(),
-                    fix_stake_deactivate,
-                );
-                if let Ok((stakers_reward, _voters_reward)) = redeemed {
-                    self.store_account(&stake_pubkey, &stake_account);
-                    vote_account_changed = true;
-
-                    if stakers_reward > 0 {
-                        rewards.push((
-                            *stake_pubkey,
-                            RewardInfo {
-                                reward_type: RewardType::Staking,
-                                lamports: stakers_reward as i64,
-                                post_balance: stake_account.lamports,
-                                commission,
-                            },
-                        ));
+                    delegations,
+                },
+            )| {
+                vote_account_rewards.insert(vote_pubkey, (vote_account, vote_state.commission, 0));
+                let vote_state = Arc::new(vote_state);
+                delegations
+                    .into_iter()
+                    .map(move |delegation| (vote_pubkey, Arc::clone(&vote_state), delegation))
```

Comment:
nit: can't just be `Arc::new(vote_state)` here without `Arc::clone()`?

Reply:
I'm not sure what you mean, we can only capture `vote_state` once and once it's `Arc` wrapped we have to clone a reference for each flat map iteration

Reply:
oh, never mind..